lib/Dialect/StaticLogic/StaticLogicOps.cpp

```diff
@@ -135,15 +135,31 @@ static LogicalResult verifyPipelineWhileOp(PipelineWhileOp op) {
   if (stagesBlock.getOperations().size() < 2)
     return op.emitOpError("stages must contain at least one stage");
 
-  // Verify the stages block contains only `staticlogic.pipeline.stage` and
-  // `staticlogic.pipeline.terminator` ops.
-  for (Operation &inner : stagesBlock)
+  int64_t lastStartTime = -1;
+  for (Operation &inner : stagesBlock) {
+    // Verify the stages block contains only `staticlogic.pipeline.stage` and
+    // `staticlogic.pipeline.terminator` ops.
     if (!isa<PipelineStageOp, PipelineTerminatorOp>(inner))
       return op.emitOpError(
                  "stages may only contain 'staticlogic.pipeline.stage' or "
                  "'staticlogic.pipeline.terminator' ops, found ")
              << inner;
 
+    // Verify the stage start times are monotonically increasing.
+    if (auto stage = dyn_cast<PipelineStageOp>(inner)) {
+      if (lastStartTime == -1) {
+        lastStartTime = stage.start();
+        continue;
+      }
+
+      if (lastStartTime >= stage.start())
```

Comment:
Just a thought; do we want to be able to support pipeline stages starting at the same time? With this, we can stall parts of a stage based on a predicate (not sure when/if this would ever be needed in the style of pipelines that we want to generate).

Reply:
That's a good question, and my current answer is no. We previously talked about doing something like this using a `when` condition attribute on the stages. But after thinking through the way we plan to use this with scheduling, @jopperm and I were talking about representing such predicated execution with an `scf.if` within a stage. This is in-line with the goal of this op capturing dataflow through a pipeline, but not the low level control (i.e. stalls in this case). I think this should be the way forward, and we should actually be removing the `when` condition (I will open an issue about this).

Reply:
https://github.com/llvm/circt/issues/2559